bokehjs/src/coffee/common/plot.coffee

```diff
@@ -196,20 +196,33 @@ class PlotView extends ContinuumView
   map_to_screen: (x, y, x_name='default', y_name='default') ->
     @frame.map_to_screen(x, y, @canvas, x_name, y_name)
 
-  update_range: (range_info) ->
+  _update_single_range: (rng, range_info, is_panning) ->
+      # Prevent range from going outside limits
+      # Also ensure that range keeps the same delta when panning
+      if rng.get('bound_lower')?
+        if rng.get('bound_lower') >= range_info['start']
+          range_info['start'] = rng.get('bound_lower')
+          if is_panning?
+            range_info['end'] = rng.get('end')
+      if rng.get('bound_upper')?
+        if rng.get('bound_upper') <= range_info['end']
+          range_info['end'] = rng.get('bound_upper')
+          if is_panning?
+            range_info['start'] = rng.get('start')
+
+      # Update the range if necessary
+      if rng.get('start') != range_info['start'] or rng.get('end') != range_info['end']
+        rng.set(range_info)
+        rng.get('callback')?.execute(@model)
+
+  update_range: (range_info, is_panning) ->
     if not range_info?
       range_info = @initial_range_info
     @pause()
     for name, rng of @frame.get('x_ranges')
-      if rng.get('start') != range_info.xrs[name]['start'] or
-          rng.get('end') != range_info.xrs[name]['end']
-        rng.set(range_info.xrs[name])
-        rng.get('callback')?.execute(@model)
+      @_update_single_range(rng, range_info.xrs[name], is_panning)
     for name, rng of @frame.get('y_ranges')
-      if rng.get('start') != range_info.yrs[name]['start'] or
-          rng.get('end') != range_info.yrs[name]['end']
-        rng.set(range_info.yrs[name])
-        rng.get('callback')?.execute(@model)
+      @_update_single_range(rng, range_info.yrs[name], is_panning)
     @unpause()
```

Comment:
This just needs to be tested with multiple ranges/axes. It might work just fine, but I'd like to see it demonstrated.